Fix: Windows drive paths misidentified as URLs

toadslop opened the pull request:
Attempts to fix this [issue](https://github.com/lycheeverse/lychee/issues/1459).

This has resulted in some failing tests on Linux. Will revise.

Edit:

Fixed Linux tests. Let me know if you have any suggestions for improvement.

mre commented on lychee-lib/src/types/input.rs:
```diff
@@ -132,7 +133,8 @@ impl Input {
         let source = if value == STDIN {
             InputSource::Stdin
-        } else if let Ok(url) = Url::parse(value) {
+        } else if let (Ok(_), Ok(url)) = (Request::builder().uri(value).body(()), Url::parse(value))
```
Your version from the issue description was a bit more readable to me, but that's probably just personal taste. Both works, so it's up to you to decide if you like to change it back.

Apart from that, can we add a comment here why this is needed? Essentially the main gist from the issue to let people know that it's a workaround for Windows.

toadslop replied:
I switched to this because the other version caused some tests to break on Linux. Apparently there are cases that RequestBuilder catches but Url doesn't, and vice versa.

mre replied:
That makes sense. It would also sounds like an important fact to document here.

mre commented:
Good job! Just a minor ask before we can merge this.

toadslop commented:
I made an adjustment to be more explicit. Relying on the behavior of RequestBuilder hides the intent a bit, so instead I explicitly say that I'm rejecting Urls that don't have an http (or https) scheme.

I also adjusted the last part of the function to handle Windows behavior better. The function originally assumed that missing a scheme could have been a mistake and adds https, but on Windows this can never happen because the separators in Windows filepaths are not valid for Urls.

Finally, I added some Windows-only tests to validate that the behavior is correct.

toadslop commented:
After the changes, though, some tests are failing that don't fail on Windows. I will need to investigate to understand the errors. They're related to something called "wayback", which I have never heard of and so I have no idea what the behavior is supposed to be.

mre commented:
Oh, the `wayback` you're referring to is short for "wayback machine", which we use to suggest archived versions in case of broken links. Unfortunately, their API is pretty flaky and lookups randomly fail, which makes these tests unreliable as well.

I suggest to ignore these particular errors for the moment. Perhaps we can even disable them entirely.

toadslop commented:
Yup, looks like you're right. I forced a test rerun and it passed.

mre commented on lychee-lib/src/types/input.rs:
```diff
@@ -132,38 +132,59 @@ impl Input {
             InputSource::Stdin
-        } else if let Ok(url) = Url::parse(value) {
-            InputSource::RemoteUrl(Box::new(url))
+        // The extra use of Request::builder is to weed out Windows filepaths with a drive specifier, which Url will erroneously parse successfully
+        // We still use Url::parse because it catches some other edge cases that Request::builder does not
+        // This could be improved with further refinement.
```
I'm guessing this comment is no longer needed? At least it looks like we don't use `Request::builder` anymore.

mre replied:
removed the comment. let me know if it's still needed.

mre commented on lychee-lib/src/types/input.rs:
```diff
@@ -132,38 +132,59 @@ impl Input {
-                    })?;
+            match Url::parse(value) {
+                // Weed out non-http schemes, including Windows drive specifiers, which will be successfully parsed by the Url crate
+                Ok(url) if url.scheme().starts_with("http") => {
```
I wonder if we should be a bit more cautious here.
Perhaps something like this?

```rust
Ok(url) if url.scheme() == "http" || url.scheme() == "https" => {
```

mre replied:
(This is just to weed out some false positives like `httpinvalid`. Maybe we could also add a test for that?)

mre replied:
done and added some tests.

mre commented:
I cleaned up the PR a bit. I think it's ready to be shipped.
Thanks for your contribution @toadslop. 😊

mre commented:
This is now released!
